10-Merge/readme.md

```diff
@@ -0,0 +1,202 @@
+
+	$ git log --all --graph --oneline --decorate
+	*   05dbdf3 (HEAD -> master) Merge branch 'bar'
+	|\  
+	| * bfb5647 (origin/bar, bar) bar
+	* | c5d665f (origin/master, origin/HEAD) fixed readme a bit
+	| | * 4909911 (origin/cherry-pick, cherry-pick) merge conflict in file
+	| | * e7f0e30 added file for cherry_pick
+	| |/  
+	|/|   
+	* | 286a273 added readme
+	|/  
+	| * c5c3da1 (origin/foo, foo) file2 for demonstration of cherry-pick
+	| * c361f2e foo
+	|/  
+	* 25f7328 file
+
+	$ cat file
+	.
+	bar
+	.
+	.
+	bar
+
+We can see that branch `bar` was merged into `master`. Now it's time to do the same with `foo`.
+
+	$ git merge foo
+	Auto-merging file
+	CONFLICT (content): Merge conflict in file
+	Automatic merge failed; fix conflicts and then commit the result.
+
+We have a conflict! We need to somehow resolve it.
+
+> You can always abort merge with `git merge --abort`
+
+Text editor way
+---------------
+
+In case of conflict, git puts changes into a file like this:
+
+	$ cat file
+	.
+	<<<<<<< HEAD
+	bar
+	.
+	.
+	bar
+	=======
+	foo
+	.
+	foo
+	.
+	>>>>>>> foo
+
+The `=======` separates two versions while `HEAD` being current branch and `foo` the merged branch (`foo`). All you have to is to modified the file into what you want it to look like, so I'll do for example this:
+
+	$ cat file
+	.
+	foobar
+	.
+	foo
+	bar
+
+After you are done with manual merge, you need to mark conflict as resolved and commit the merge:
+
+	$ git add file
+	$ git commit -m "Merge branch 'foo'"
+	[master b702845] Merge branch 'foo'
+
+Let's check it out:
+
+	$ git log --all --graph --oneline --decorate
+	*   b702845 (HEAD -> master) Merge branch 'foo'
+	|\  
+	| * c5c3da1 (origin/foo, foo) file2 for demonstration of cherry-pick
+	| * c361f2e foo
+	* |   05dbdf3 Merge branch 'bar'
+	|\ \  
+	| * | bfb5647 (origin/bar, bar) bar
+	| |/  
+	* | c5d665f (origin/master, origin/HEAD) fixed readme a bit
+	| | * 4909911 (origin/cherry-pick, cherry-pick) merge conflict in file
+	| | * e7f0e30 added file for cherry_pick
+	| |/  
+	|/|   
+	* | 286a273 added readme
+	|/  
+	* 25f7328 file
+
+We now have merge both `bar` and `foo` into the master, in the process we resolved conflict between them.
+
+Mergetool way
+-------------
+
+We throw away the last commit (the merge of `foo`), so we can do it again, this time using merge tool.
+
+	$ git reset --hard HEAD~1
+	HEAD is now at 05dbdf3 Merge branch 'bar'
+
+> `git reset --hard` basically throws away everything after specified commit and `HEAD~1` is shorthand for `one-before-last` commit
+
+Now do the merge
+
+	$ git merge foo
+	Auto-merging file
+	CONFLICT (content): Merge conflict in file
+	Automatic merge failed; fix conflicts and then commit the result.
+	$ git mergetool
+
+Now resolve conflict in your favourite mergetool you configured and commit the result.
+
+Cherry-pick
```

Comment:
I think it can stay in here, git-cherry-pick is relatively short and imo strongly related to merges